[MultiKueue] integration tests fail occasionally

/kind flake

**What happened**:

Failure on the unrelated branch: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kueue/2362/pull-kueue-test-integration-main/1798319515201703936

**What you expected to happen**:

No failure

**How to reproduce it (as minimally and precisely as possible)**:

Probably repeat the build multiple times.

**Anything else we need to know?**:

The report from the build log:
```
  In [ReportAfterSuite] at: autogenerated by Ginkgo @ 06/05/24 11:50:42.911
------------------------------
Ginkgo timed out waiting for all parallel procs to report back
Test suite: multikueue (./test/integration/multikueue)
This occurs if a parallel process exits before it reports its results to the
Ginkgo CLI.  The CLI will now print out all the stdout/stderr output it's
collected from the running processes.  However you may not see anything useful
in these logs because the individual test processes usually intercept output to
stdout/stderr in order to capture it in the spec reports.
You may want to try rerunning your test suite with
--output-interceptor-mode=none to see additional output here and debug your
suite.
  
Output from proc 1:
  --- FAIL: TestMultiKueue (31.90s)
  FAIL
Output from proc 2:
Output from proc 3:
  PASS
Output from proc 4:
  PASS
** End **Could not open /logs/artifacts/test_integration_multikueue_junit.xml:
open /logs/artifacts/test_integration_multikueue_junit.xml: no such file or directory
Ginkgo ran 20 suites in 6m39.250168558s
There were failures detected in the following suites:
  multikueue ./test/integration/multikueue
Test Suite Failed
make: *** [Makefile-test.mk:68: test-integration] Error 1
```

--- a/pkg/controller/admissionchecks/multikueue/jobset_adapter_test.go
+++ b/pkg/controller/admissionchecks/multikueue/jobset_adapter_test.go
@@ -342,6 +342,7 @@ func TestWlReconcileJobset(t *testing.T) {
 
 			w1remoteClient := newRemoteClient(managerClient, nil, nil, defaultOrigin, "")
 			w1remoteClient.client = worker1Client
+			w1remoteClient.connecting.Store(false)
 
 			cRec.remoteClients["worker1"] = w1remoteClient
 

--- a/pkg/controller/admissionchecks/multikueue/multikueuecluster.go
+++ b/pkg/controller/admissionchecks/multikueue/multikueuecluster.go
@@ -81,7 +81,7 @@ type remoteClient struct {
 	kubeconfig   []byte
 	origin       string
 
-	pendingReconnect   atomic.Bool
+	connecting         atomic.Bool
 	failedConnAttempts uint
 
 	// For unit testing only. There is now need of creating fully functional remote clients in the unit tests
@@ -98,6 +98,7 @@ func newRemoteClient(localClient client.Client, wlUpdateCh, watchEndedCh chan<- 
 		localClient:  localClient,
 		origin:       origin,
 	}
+	rc.connecting.Store(true)
 	return rc
 }
 
@@ -139,7 +140,7 @@ func (*workloadKueueWatcher) GetWorkloadKey(o runtime.Object) (types.NamespacedN
 // If the encountered error is not permanent the duration after which a retry should be done is returned.
 func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) (*time.Duration, error) {
 	configChanged := !equality.Semantic.DeepEqual(kubeconfig, rc.kubeconfig)
-	if !configChanged && !rc.pendingReconnect.Load() {
+	if !configChanged && !rc.connecting.Load() {
 		return nil, nil
 	}
 
@@ -183,7 +184,7 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) (
 		}
 	}
 
-	rc.pendingReconnect.Store(false)
+	rc.connecting.Store(false)
 	rc.failedConnAttempts = 0
 	return nil, nil
 }
@@ -218,9 +219,9 @@ func (rc *remoteClient) startWatcher(ctx context.Context, kind string, w multiKu
 		log.V(2).Info("Watch ended", "ctxErr", ctx.Err())
 		// If the context is not yet Done , queue a reconcile to attempt reconnection
 		if ctx.Err() == nil {
-			oldReconnect := rc.pendingReconnect.Swap(true)
+			oldConnecting := rc.connecting.Swap(true)
 			// reconnect if this is the first watch failing.
-			if !oldReconnect {
+			if !oldConnecting {
 				log.V(2).Info("Queue reconcile for reconnect", "cluster", rc.clusterName)
 				rc.queueWatchEndedEvent(ctx)
 			}
@@ -258,6 +259,12 @@ func (rc *remoteClient) queueWatchEndedEvent(ctx context.Context) {
 // is owned by a job, also delete the job.
 func (rc *remoteClient) runGC(ctx context.Context) {
 	log := ctrl.LoggerFrom(ctx)
+
+	if rc.connecting.Load() {
+		log.V(5).Info("Skip disconnected client")
+		return
+	}
+
 	lst := &kueue.WorkloadList{}
 	err := rc.client.List(ctx, lst, client.MatchingLabels{kueuealpha.MultiKueueOriginLabel: rc.origin})
 	if err != nil {

--- a/pkg/controller/admissionchecks/multikueue/multikueuecluster_test.go
+++ b/pkg/controller/admissionchecks/multikueue/multikueuecluster_test.go
@@ -76,7 +76,7 @@ func newTestClient(config string, watchCancel func()) *remoteClient {
 
 func setReconnectState(rc *remoteClient, a uint) *remoteClient {
 	rc.failedConnAttempts = a
-	rc.pendingReconnect.Store(true)
+	rc.connecting.Store(true)
 	return rc
 }
 
@@ -535,6 +535,7 @@ func TestRemoteClientGC(t *testing.T) {
 
 			w1remoteClient := newRemoteClient(managerClient, nil, nil, defaultOrigin, "")
 			w1remoteClient.client = worker1Client
+			w1remoteClient.connecting.Store(false)
 
 			w1remoteClient.runGC(ctx)
 

--- a/pkg/controller/admissionchecks/multikueue/workload.go
+++ b/pkg/controller/admissionchecks/multikueue/workload.go
@@ -247,7 +247,7 @@ func (w *wlReconciler) remoteClientsForAC(ctx context.Context, acName string) (m
 	for _, clusterName := range cfg.Spec.Clusters {
 		if client, found := w.clusters.controllerFor(clusterName); found {
 			// Skip the client if its reconnect is ongoing.
-			if !client.pendingReconnect.Load() {
+			if !client.connecting.Load() {
 				clients[clusterName] = client
 			}
 		}

--- a/pkg/controller/admissionchecks/multikueue/workload_test.go
+++ b/pkg/controller/admissionchecks/multikueue/workload_test.go
@@ -1012,6 +1012,7 @@ func TestWlReconcile(t *testing.T) {
 
 			w1remoteClient := newRemoteClient(managerClient, nil, nil, defaultOrigin, "")
 			w1remoteClient.client = worker1Client
+			w1remoteClient.connecting.Store(false)
 			cRec.remoteClients["worker1"] = w1remoteClient
 
 			var worker2Client client.WithWatch
@@ -1042,8 +1043,8 @@ func TestWlReconcile(t *testing.T) {
 
 				w2remoteClient := newRemoteClient(managerClient, nil, nil, defaultOrigin, "")
 				w2remoteClient.client = worker2Client
-				if tc.worker2Reconnecting {
-					w2remoteClient.pendingReconnect.Store(true)
+				if !tc.worker2Reconnecting {
+					w2remoteClient.connecting.Store(false)
 				}
 				cRec.remoteClients["worker2"] = w2remoteClient
 			}

